### PR TITLE
Add new CaloL1 offline DQM module and configuration

### PR DESCRIPTION
**Pull request by aloeliger**
#### PR description:

A new CaloL1 DQM offline module in "Global" format has been introduced to comply with the concurrent LS processing introduced for the offline DQM as a part of #25090. This offline module is largely a duplicate of the online version of the L1TStage2CaloLayer1 module, but with 7 monitoring elements removed, and changed to be a DQMGlobalEDAnalyzer instead of DQMOneEDAnalyzer. Monitoring elements have been moved to a dedicated CaloL1Information namespace and data structure as part of the DQMGlobalEDAnalyzer template. A duplicate of the online _cfi.py has been created for the offline module, and the overall offline L1DQM sequence L1TriggerDqmOffline_cff.py has been adjusted to add this module into the offline workflow.

#### PR validation:

All code compiles, and I have checked that a min bias data workflow reproduces the same histograms between the old online module and new offline module. I have also checked runTheMatrix test workflows and any failures seem due to failing to find or open remote files. As well, the PR seems to pass the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html).

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This PR is not a backport

### CODE REVIEW

**Comment by silviodonato** [comment]
please test

File: DQMOffline/L1Trigger/interface/L1TStage2CaloLayer1Offline.h [hunk 1]
```diff
@@ -0,0 +1,198 @@
+#ifndef L1TStage2CaloLayer1Offline_H
+#define L1TStage2CaloLayer1Offline_H
+
+//change from "One" to "Global"
```
Comment (silviodonato) [inline DQMOffline/L1Trigger/interface/L1TStage2CaloLayer1Offline.h:4]: You can remove this kind of comments. Perhaps you can write a general comment saying that this is almost a duplicate of the online version of the L1TStage2CaloLayer1

**Comment by silviodonato** [comment]
@aloeliger  Could you explain why you are adding a new modules instead of modifying L1TStage2CaloLayer1?

**Comment by aloeliger** [comment]
@silviodonato There are seven luminosity block transition based monitoring elements that are not compatible with the change from `DQMOneEDAnalyzer<edm::one::WatchLuminosityBlocks>` to `DQMGlobalEDAnalzyer<>` these are:

1. `last20Mismatches_`: A monitoring element that was read back at the beginning of luminosity blocks detailing the last 20 mismatches by type (ECAL/HCAL and fine grain and feature versus other) and run, luminosity section and event. 
2. `maxEvtLinkErrorsByLumi(E/H)CAL_` and `maxEvtLinkErrorsByLumi_`: these 3 are monitoring elements that track the max number of link errors seen in an event for ECAL, HCAL or both for a given luminosity section. These were filled at the end of a luminosity block, and the information needed to do this would need to be shared across streams
3. `maxEvtMismatchByLumi(E/H)CAL_` and `maxEvtMismatchByLumi:_` These 3 are monitoring elements similar to the above, but for trigger mismatches.

The CaloL1 team was reluctant to change monitoring elements used for online monitoring, so for a compromise I left online monitoring unchanged and introduced a slightly changed duplicate for a multithreaded use for offline DQM.

File: DQMOffline/L1Trigger/interface/L1TStage2CaloLayer1Offline.h (hunk 1 above)
Comment (aloeliger) [reply DQMOffline/L1Trigger/interface/L1TStage2CaloLayer1Offline.h:4]: I can go change this. Will this require the code checks and tests be restarted?

**Comment by aloeliger** [comment]
I am not sure I understand the nature of the test failure here:

> RAW2DIGI:siPixelDigis,VALIDATION
We have determined that this is real data (if not, rerun cmsDriver.py with --mc)
entry /store/data/Run2018A/EGamma/RAW/v1/000/315/489/00000/004D960A-EA4C-E811-A908-FA163ED1F481.root
Step: RAW2DIGI Spec: ['siPixelDigis']
Step: VALIDATION Spec: 
 in preparing validation
Traceback (most recent call last):
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/bin/slc7_amd64_gcc820/cmsDriver.py", line 56, in <module>
    run()
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/bin/slc7_amd64_gcc820/cmsDriver.py", line 28, in run
    configBuilder.prepare()
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/Configuration/Applications/ConfigBuilder.py", line 2135, in prepare
    self.addStandardSequences()
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/Configuration/Applications/ConfigBuilder.py", line 757, in addStandardSequences
    getattr(self,"prepare_"+stepName)(sequence = getattr(self,stepName+"DefaultSeq"))
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/Configuration/Applications/ConfigBuilder.py", line 1773, in prepare_VALIDATION
    self.loadDefaultOrSpecifiedCFF(sequence,self.VALIDATIONDefaultCFF)
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/Configuration/Applications/ConfigBuilder.py", line 1198, in loadDefaultOrSpecifiedCFF
    l=self.loadAndRemember(defaultCFF)
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/Configuration/Applications/ConfigBuilder.py", line 325, in loadAndRemember
    self.process.load(includeFile)
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/FWCore/ParameterSet/Config.py", line 684, in load
    self.extend(sys.modules[moduleName])
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/FWCore/ParameterSet/Config.py", line 731, in extend
    self.__setattr__(name, task)
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/FWCore/ParameterSet/Config.py", line 507, in __setattr__
    newValue._place(name,self)
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/FWCore/ParameterSet/SequenceTypes.py", line 1441, in _place
    proc._placeTask(name,self)
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/FWCore/ParameterSet/Config.py", line 650, in _placeTask
    self._validateTask(task, name)
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-10-05-2300/python/FWCore/ParameterSet/Config.py", line 884, in _validateTask
    raise RuntimeError("An entry in task " + label + ' has not been attached to the process')
RuntimeError: An entry in task recoJetsHIpostAODTask has not been attached to the process
Analyzing 50 seqs...
Task failed.
Traceback (most recent call last):
  File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-05-2300/bin/slc7_amd64_gcc820/cmsswSequenceInfo.py", line 535, in <module>
    processseqs(seqset)
  File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-05-2300/bin/slc7_amd64_gcc820/cmsswSequenceInfo.py", line 293, in processseqs
    for it in t.get(): # should only be one
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/external/python3/3.8.2-bcolbf/lib/python3.8/multiprocessing/pool.py", line 768, in get
    raise self._value
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/external/python3/3.8.2-bcolbf/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/external/python3/3.8.2-bcolbf/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-05-2300/bin/slc7_amd64_gcc820/cmsswSequenceInfo.py", line 277, in <lambda>
    tasks = [stp.map_async(lambda seq: (seq, inspectsequence(seq)), [seq]) for seq in seqs]
  File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-05-2300/bin/slc7_amd64_gcc820/cmsswSequenceInfo.py", line 70, in inspectsequence
    subprocess.check_call(driverargs, cwd=wd, stdout=2) # 2: STDERR
  File "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/external/python3/3.8.2-bcolbf/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmsDriver.py', 'step3', '--conditions', 'auto:run2_data', '-s', 'RAW2DIGI:siPixelDigis,VALIDATION', '--process', 'DUMMY', '--eventcontent', 'DQM', '--scenario', 'HeavyIons', '--datatier', 'DQMIO', '--customise_commands', 'process.Tracer = cms.Service("Tracer")', '--filein', '/store/data/Run2018A/EGamma/RAW/v1/000/315/489/00000/004D960A-EA4C-E811-A908-FA163ED1F481.root', '-n', '0', '--python_filename', 'cmssw_cfg.py', '--no_exec']' returned non-zero exit status 1.

> 

> ---> test TestDQMOfflineConfiguration0 had ERRORS

>  

> ^^^^ End Test TestDQMOfflineConfiguration0 ^^^^

If I am reading this correctly the failure in this step is before the DQM step occurs, and is in a silicon pixel DIGI step. It also seems like the RuntimeError that occurs mentions jet reconstruction and heavy ions. I am not sure how my changes could have affected these processes.

**Comment by jfernan2** [comment]
@aloeliger  the Unit Tests errors are unrelated to your PR, see https://github.com/cms-sw/cmssw/issues/31669  for details
Please consider the comments I made you by email for this PR

**Comment by jfernan2** [comment]
> @aloeliger the Unit Tests errors are unrelated to your PR, see #31669 for details
> Please consider the comments I made you by email for this PR
OK, I see now that @silviodonato made the same comments here in github

File: DQMOffline/L1Trigger/interface/L1TStage2CaloLayer1Offline.h (hunk 1 above)
Comment (silviodonato) [reply DQMOffline/L1Trigger/interface/L1TStage2CaloLayer1Offline.h:4]: Yes, when you push the changes to your repository, the code-checks will start again automatically.
Please remove all comments similar to this one
Example:
`//#include "DQMServices/Core/interface/DQMOneEDAnalyzer.h" `
or
`//Elements removed from online version of module. [...]`

**Comment by aloeliger** [comment]
The code commenting should now be somewhat cleaned-up

**Comment by silviodonato** [comment]
please test with  #31674, #31309

**Comment by jfernan2** [comment]
@aloeliger  I notice now that the reportSummary Map becomes orange for CaloLayer 1 and HCAL/ECAL TPG [1] due to this PR since I guess it was based on the MEs you are removing.
I would be good to modify the report so that it does not check for those plots
See what I mean here:
[1] https://tinyurl.com/y52ltpuo

**Comment by aloeliger** [comment]
@jfernan2 I hadn't considered this.

The tests for that summary are configured here:

https://github.com/cms-sw/cmssw/blob/50769d0804b2a858c7db42c85e77827bba2b4561/DQM/L1TMonitorClient/data/L1TStage2CaloLayer1QualityTests.xml#L1-L38

The only tests considered are for monitoring elements I removed.

I think there are a few possible solutions here:

1. I could make an offline version of these and change these tests from the maximum number of link errors or mismatches per bunch crossing, into just considering the absolute number for each luminosity section.

This of course makes things a little more problematic because it's another unsynchronized element with the online version of the DQM, and I don't really have a good way to judge what the thresholds should be. The good news is that it would be pretty quick.

2. I could go back and investigate regaining the functionality of these monitoring elements in the global context

I've been thinking about this idea yesterday and today, and there might be a way to introduce some kind of dedicated data storage class that is a part of the `DQMGlobalEDAnalyzer<>` RunCache that processes and stores information about the lumi sections across streams (as I think about it, even a `std::map` might do) because the information needed about the lumisection is not too complicated. This would be good because we could get rid of a separate module for offline, and go back to just one module in use for online and offline both. The downside to this is that I would need to spend some more time working this out again, and then making sure that online monitoring has not changed.

**Comment by jfernan2** [comment]
Hi @aloeliger 
The last solution you propose seems the best, but it needs some work of course

**Comment by silviodonato** [comment]
@aloeliger do you have any news?

**Comment by aloeliger** [comment]
Hi @silviodonato, I am looking at a change that would supersede this one and not require separate offline components and configurations, and would get the quality tests back. 

Hopefully this can fulfill CaloL1's online needs and threaded offline needs by using a structure with maps of lumi-sections and the 'max*' information that the quality tests depend on. This can be updated event by event and will be read out all at once at the end of a run instead of the end of a lumi-section and should maintain all the current information used by CaloL1 while simultaneously being "Global". 

These module changes compile, but I am working through a series of run time segfaults. I am afraid I am preparing for a talk at a conference right now, so updates may be a little slow.

**Comment by silviodonato** [comment]
The previous test was done just to check that the framework does not complain anymore about  `The following modules require synchronizing on LuminosityBlock boundaries` of `L1TStage2CaloLayer1`. The test was successful.

**Comment by silviodonato** [comment]
@aloeliger @cms-sw/dqm-l2 @cms-sw/l1-l2 what is the status of this PR? Is it ready to be merged?

**Comment by aloeliger** [comment]
> @aloeliger I notice now that the reportSummary Map becomes orange for CaloLayer 1 and HCAL/ECAL TPG [1] due to this PR since I guess it was based on the MEs you are removing.
> I would be good to modify the report so that it does not check for those plots
> See what I mean here:
> [1] https://tinyurl.com/y52ltpuo

> @aloeliger @cms-sw/dqm-l2 @cms-sw/l1-l2 what is the status of this PR? Is it ready to be merged?

@silviodonato, @jfernan2, apologies again for my schedule, but now that Higgs2020 is done, I have talked to the CaloL1 team about a new fix maintains the summary reports that were broken. I have opened a new PR #32004 with changes exclusively to the online CaloL1 DQM that does not require this separate offline module or any changes to any configurations. At this point, I think this PR is no longer necessary, and could be safely closed in favor of #32004.

**Comment by silviodonato** [comment]
moved to https://github.com/cms-sw/cmssw/pull/32004

**Pull request closed without merging**